### PR TITLE
add file upload

### PR DESCRIPTION
**Pull request by Lijia-Li**
- add file name in return page
- reconfigure front page by adding upload file choice 
- add file upload and use the uploaded files as search base

front_page_change_workflow

### CODE REVIEW

**Comment by Lijia-Li** [comment]
@justinnhli @crzeller11

File: ReCord_frontend.py [hunk 1]
```diff
@@ -1,10 +1,21 @@
...
+UPLOAD_FOLDER = './uploads/'
+ALLOWED_EXTENSIONS = {'xml'}
 
+# initiate the app
 app = Flask(__name__)
+app.config['UPLOAD_FOLDER'] = UPLOAD_FOLDER
+app.secret_key = '\x82\xebT\x17\x07\xbbx\xd9\xe1dxR\x11\x8b\x0ci\xe1\xb7\xa8\x97\n\xd6\x01\x99'
+
+
+def allowed_file(filename):
```
Comment (justinnhli) [inline ReCord_frontend.py]: docstring needed

File: ReCord_frontend.py [hunk 2]
```diff
@@ -20,16 +31,102 @@ def my_form_post():
...
+        # elif tag == 'Slur/Ligatures':
+        # elif tag == 'Ornaments':
+        # elif tag == 'Notes':
+        # elif tag == 'Accidental':
+
+        return render_template('ReChord_result.html', result=result)
+
+
+@app.route('/uploads/<filename>')
+def uploaded_file(filename):
```
Comment (justinnhli) [inline ReCord_frontend.py]: docstring needed

File: ReCord_frontend.py [hunk 3]
```diff
@@ -20,16 +31,102 @@ def my_form_post():
...
+    return send_from_directory(app.config['UPLOAD_FOLDER'],
+                               filename)
+
+
+def upload_file(name_tag):
```
Comment (justinnhli) [inline ReCord_frontend.py]: docstring needed

File: ReCord_frontend.py [hunk 4]
```diff
@@ -20,16 +31,102 @@ def my_form_post():
     """the view function which return the result page by using the input pass to the back end
```
Comment (justinnhli) [inline ReCord_frontend.py]: Everything below should be broken down into their own functions, intead of having a super long function here.

**Review by justinnhli — CHANGES_REQUESTED** [review]

File: ReCord_frontend.py [hunk 5]
```diff
@@ -1,10 +1,21 @@
-from search import prepare_tree, search, find_artic, get_measure
+from search import *
 from lxml import etree
-from flask import Flask, request, render_template, flash
+from flask import Flask, request, render_template, flash, redirect, url_for, send_from_directory
 from io import BytesIO
+from werkzeug.utils import secure_filename
 
+UPLOAD_FOLDER = './uploads/'
+ALLOWED_EXTENSIONS = {'xml'}
```
Comment (justinnhli) [inline ReCord_frontend.py]: I thought you wanted to allow ".mei" files?

**Comment by justinnhli** [comment]
Can you also give me a brief overview of how this uploading is supposed to work?

Also, in general, look at the lint output for rechord_search.py and fix them.

**Comment by Lijia-Li** [comment]
@justinnhli Done fixing. Let me know if there are more things need to be fix. Some linting issue need bigger changes that I prefer to save till later.

File: ReCord_frontend.py [hunk 6]
```diff
@@ -1,35 +1,129 @@
...
 
+# initiate the app
+App = Flask(__name__)
+App.config['UPLOAD_FOLDER'] = UPLOAD_FOLDER
+App.secret_key = '\x82\xebT\x17\x07\xbbx\xd9\xe1dxR\x11\x8b\x0ci\xe1\xb7\xa8\x97\n\xd6\x01\x99'
 
-app = Flask(__name__)
 
+def allowed_file(filename):
+    """check the file name to avoid possible hack"""
```
Comment (justinnhli) [inline ReCord_frontend.py]: Need argument and return value description in docstring.

File: ReCord_frontend.py [hunk 7]
```diff
@@ -1,35 +1,129 @@
...
+        tag = request.form['term']
+        para = request.form['parameter']
+        tree, _ = prepare_tree('database/Chopin.xml')
+        return search_terms(tag, para, tree)
+    return
+
+
+@App.route('/uploads/<filename>')
+def uploaded_file(filename):
+    """upload file and add to upload folder"""
```
Comment (justinnhli) [inline ReCord_frontend.py]: Arguments and return value description?

File: ReCord_frontend.py [hunk 8]
```diff
@@ -1,35 +1,129 @@
...
+    return send_from_directory(App.config['UPLOAD_FOLDER'],
+                               filename)
+
+
+# Helper function
+
+
+def search_snippet(snippet, tree):
+    """search the snippet from the given database"""
```
Comment (justinnhli) [inline ReCord_frontend.py]: Arguments and return value description?

File: ReCord_frontend.py [hunk 9]
```diff
@@ -1,35 +1,129 @@
...
+    input_root = input_xml.getroot()
+
+    snippet_measure = search(input_root, tree)
+    title = get_title(tree)
+    creator = get_creator(tree)
+    return render_template('ReChord_result.html', results=snippet_measure, title=title, creator=creator)
+
+
+def search_terms(tag, para, tree):
+    """ search terms in the database"""
```
Comment (justinnhli) [inline ReCord_frontend.py]: Arguments and return value description?

File: ReCord_frontend.py [hunk 10]
```diff
@@ -1,35 +1,129 @@
...
+    # elif tag == 'Hairpin':
+    # elif tag == 'Slur/Ligatures':
+    # elif tag == 'Ornaments':
+    # elif tag == 'Notes':
+    # elif tag == 'Accidental':
+    return render_template('ReChord_result.html', result=result)
+
+
+def upload_file(name_tag):
+    """pass the upload file and store it in uploads folder"""
```
Comment (justinnhli) [inline ReCord_frontend.py]: Arguments and return value description?

File: ReCord_frontend.py [hunk 11]
```diff
@@ -1,35 +1,129 @@
...
 from io import BytesIO
+from werkzeug.utils import secure_filename
+from lxml import etree
+from search import search, prepare_tree, get_title, get_creator, find_artic, find_expressive_term, os
+
+UPLOAD_FOLDER = './uploads/'
+ALLOWED_EXTENSIONS = {'xml', 'mei'}
 
+# initiate the app
+App = Flask(__name__)
```
Comment (justinnhli) [inline ReCord_frontend.py]: Do this to make pylint stop complaining about the variable name:

```python
App = Flask(__name__) # pylint: disable=invalid-name
```

**Review by justinnhli — CHANGES_REQUESTED** [review]
almost there!

File: ReCord_frontend.py (hunk 11 above)
Comment (Lijia-Li) [reply ReCord_frontend.py]: Thank you.

**Comment by Lijia-Li** [comment]
@justinnhli added arguments in docstring and fixed some more linting problems.

**Pull request merged**